src/lib.rs

```diff
@@ -1,7 +1,1242 @@
+                    .mul(expsk.ask, params),
+                nk: params
+                    .generator(FixedGenerators::ProofGenerationKey)
+                    .mul(expsk.nsk, params),
+            },
+            ovk: expsk.ovk,
+        }
+    }
+
+    fn derive_child(&self, i_l: &[u8], params: &E::Params) -> Self {
+        let i_ask = E::Fs::to_uniform(prf_expand(i_l, &[0x13]).as_bytes());
+        let i_nsk = E::Fs::to_uniform(prf_expand(i_l, &[0x14]).as_bytes());
+        let ak = params
+            .generator(FixedGenerators::SpendingKeyGenerator)
+            .mul(i_ask, params)
+            .add(&self.vk.ak, params);
+        let nk = params
+            .generator(FixedGenerators::ProofGenerationKey)
+            .mul(i_nsk, params)
+            .add(&self.vk.nk, params);
+
+        FullViewingKey {
+            vk: ViewingKey { ak, nk },
+            ovk: self.ovk.derive_child(i_l),
+        }
+    }
+
+    pub fn read<R: Read>(mut reader: R, params: &E::Params) -> io::Result<Self> {
+        let ak = edwards::Point::<E, Unknown>::read(&mut reader, params)?;
+        let ak = match ak.as_prime_order(params) {
+            Some(p) => p,
+            None => {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "ak not of prime order",
+                ))
+            }
+        };
+
+        let nk = edwards::Point::<E, Unknown>::read(&mut reader, params)?;
+        let nk = match nk.as_prime_order(params) {
+            Some(p) => p,
+            None => {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "nk not of prime order",
+                ))
+            }
+        };
+
+        let mut ovk = [0; 32];
+        reader.read_exact(&mut ovk)?;
+
+        Ok(FullViewingKey {
+            vk: ViewingKey { ak, nk },
+            ovk: OutgoingViewingKey(ovk),
+        })
+    }
+
+    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        self.vk.ak.write(&mut writer)?;
+        self.vk.nk.write(&mut writer)?;
+        writer.write_all(&self.ovk.0)?;
+
+        Ok(())
+    }
+
+    fn to_bytes(&self) -> [u8; 96] {
+        let mut result = [0u8; 96];
+        self.write(&mut result[..])
+            .expect("should be able to serialize a FullViewingKey");
+        result
+    }
+}
+
+// ZIP 32 structures
+
+/// A Sapling full viewing key fingerprint
+struct FVKFingerprint([u8; 32]);
+
+impl<'a, E: JubjubEngine> From<&'a FullViewingKey<E>> for FVKFingerprint {
+    fn from(fvk: &FullViewingKey<E>) -> Self {
+        let mut h = Blake2b::with_params(32, &[], &[], ZIP32_SAPLING_FVFP_PERSONALIZATION);
+        h.update(&fvk.to_bytes());
+        let mut fvfp = [0u8; 32];
+        fvfp.copy_from_slice(h.finalize().as_bytes());
+        FVKFingerprint(fvfp)
+    }
+}
+
+/// A Sapling full viewing key tag
+#[derive(Clone, Copy, Debug, PartialEq)]
+struct FVKTag([u8; 4]);
```

Comment:
We have several structs in this file which contain only one field. I am wondering if this code would be easier to follow/more readable without these.

For example instead of having the ```FVKFingerprint``` and ```FVKTag```, ```FullViewingKey<E>``` could be written as follows:
```
/// A Sapling full viewing key
pub struct FullViewingKey<E: JubjubEngine> {
    vk: ViewingKey<E>,
    ovk: [u8; 32],
}

impl<E: JubjubEngine> FullViewingKey<E> {
    fn fingerprint(&self) -> [u8; 32] {
        let mut h = Blake2b::with_params(32, &[], &[], ZIP32_SAPLING_FVFP_PERSONALIZATION);
        h.update(&self.to_bytes());
        let mut fvfp = [0u8; 32];
        fvfp.copy_from_slice(h.finalize().as_bytes());
        fvfp
    }

    fn tag(&self) -> [u8; 4] {
        let mut tag = [0u8; 4];
        tag.copy_from_slice(&self.fingerprint()[..4]);
        tag
    }
}
```
and then 
```assert_eq!(FVKFingerprint::from(&xfvk.fvk).0, tv.fp);``` would become ```assert_eq!(xfvk.fvk.fingerprint(), tv.fp);```

Reply:
Replacing the `From<>` impls with plain functions might clean things up. But we should keep the structs themselves, as they provide type-safety. A struct containing a single field is called a newtype in Rust, and is a zero-cost abstraction (once compiled, the outer struct behaves exactly as if it was the inner type).